# KHOA HỌC DỮ LIỆU & ỨNG DỤNG
### GVHD: TRẦN TRUNG KIÊN
### THỰC HIỆN: Nhóm 12
### DANH SÁCH THÀNH VIÊN
  1. 1712667 - Lê Hữu Phúc
  2. 18120167 - Nguyễn Viết Dũng
  3. 18120579 - Đặng Minh Thọ
  4. 18120584 - Phạm Đình Thục

# 1. GIỚI THIỆU ĐỀ TÀI




- **Tên cuộc thi:** Recruit Restaurant Visitor Forecasting
- **Mô tả tổng quan:** Dự đoán nhà hàng sẽ nhận được bao nhiêu khách trong tương lai
- **Prize:** $25000
- **Participants:** 2148 teams
- **Link cuộc thi:** https://www.kaggle.com/c/recruit-restaurant-visitor-forecasting
- **Link `TOP 8` Solution:** https://github.com/MaxHalford/kaggle-recruit-restaurant


# 2. SƠ LƯỢC VỀ DỮ LIỆU


Dữ liệu đến từ hai trang web riêng biệt: 
  - Hot Pepper Gourmet (hpg): tại đây người dùng có thể tìm kiếm nhà hàng và đặt chỗ trực tuyến 
  - AirREGI / Restaurant Board (air): hệ thống kiểm soát đặt chỗ và tính tiền
  
Được thu thập từ năm 2016 đến tháng 4 năm 2017 
- AIR: 829 restaurants 
- HPG: 4690 restaurants

Dữ liệu thu thập bao gồm các thông tin liên quan đến nhà hàng (id, latitude, longitude, ...), các thông tin về số lượng khách đặt chỗ tại các nhà hàng và các thông tin liên quan đến thời tiết, ngày lễ.

# 3. IMPORT CÁC THƯ VIỆN CẦN THIẾT

In [ ]:
# Mount Drive
from google.colab import drive
import os
drive.mount('/content/drive')
os.chdir("/content/drive/My Drive/Colab Notebooks/KHDL")
!ls


import pandas as pd
import numpy as np

Mounted at /content/drive
data


# 4. KHÁM PHÁ DỮ LIỆU


---
## 4.1 File `air_visit_data.csv`

Trong file `air_visit_data.csv` liên quan về dữ liệu khách theo ngày: Trong file biểu diễn dữ liệu theo ngày (những ngày không có khách sẽ không được thu thập).

**Giải pháp:**

 -> Ta cần phải lấy mẫu theo ngày thay vì theo lượt truy cập.
 
  Do đó, những ngày không có khách thì lượt truy cập sẽ là 0. Có ích cho việc tính toán luân phiên dựa vào thời gian

In [ ]:
air_visit = pd.read_csv('data/air_visit_data.csv')

air_visit.head()

air_store_id  visit_date  visitors
0  air_ba937bf13d40fb24  2016-01-13        25
1  air_ba937bf13d40fb24  2016-01-14        32
2  air_ba937bf13d40fb24  2016-01-15        29
3  air_ba937bf13d40fb24  2016-01-16        22
4  air_ba937bf13d40fb24  2016-01-18         6

* Sắp xếp dữ liệu theo ngày và thêm vào những ngày không được thu thập (số lượng khách bằng NaN)






In [ ]:
air_visit.index = pd.to_datetime(air_visit['visit_date'])
air_visit = air_visit.groupby('air_store_id').apply(lambda g: g['visitors'].resample('1d').sum()).reset_index()
air_visit.replace(0, np.nan, inplace=True)
air_visit.head()

air_store_id visit_date  visitors
0  air_00a91d42b08b08d9 2016-07-01      35.0
1  air_00a91d42b08b08d9 2016-07-02       9.0
2  air_00a91d42b08b08d9 2016-07-03       NaN
3  air_00a91d42b08b08d9 2016-07-04      20.0
4  air_00a91d42b08b08d9 2016-07-05      25.0

* Gán cờ `was_nil` cho những ngày không có khách

In [ ]:
air_visit['visit_date'] = air_visit['visit_date'].dt.strftime('%Y-%m-%d')
air_visit['was_nil'] = air_visit['visitors'].isnull()
air_visit['visitors'].fillna(0, inplace=True)

air_visit.head()

air_store_id  visit_date  visitors  was_nil
0  air_00a91d42b08b08d9  2016-07-01      35.0    False
1  air_00a91d42b08b08d9  2016-07-02       9.0    False
2  air_00a91d42b08b08d9  2016-07-03       0.0     True
3  air_00a91d42b08b08d9  2016-07-04      20.0    False
4  air_00a91d42b08b08d9  2016-07-05      25.0    False

---
## 4.2 File `date_info.csv`


Trong file `date_info.csv` liên quan đến dữ liệu các ngày lễ.

Bao gồm: Thứ, ngày, tháng, năm và cờ để xác định có phải ngày lễ hay không?

**Giải pháp:** 
 Ta thêm hai tính năng cho biết ngày hôm trước và ngày hôm sau là ngày lễ hay không?

 -> Điều này giúp ta biết được ngày đó là ngày lễ đơn hay là kỳ nghỉ lễ dài ngày

In [ ]:
date_info = pd.read_csv('data/date_info.csv')

date_info.head()

calendar_date day_of_week  holiday_flg
0    2016-01-01      Friday            1
1    2016-01-02    Saturday            1
2    2016-01-03      Sunday            1
3    2016-01-04      Monday            0
4    2016-01-05     Tuesday            0

In [ ]:
date_info.rename(columns={'holiday_flg': 'is_holiday', 'calendar_date': 'visit_date'}, inplace=True)
date_info['prev_day_is_holiday'] = date_info['is_holiday'].shift().fillna(0)
date_info['next_day_is_holiday'] = date_info['is_holiday'].shift(-1).fillna(0)

date_info.head()

visit_date day_of_week  is_holiday  prev_day_is_holiday  next_day_is_holiday
0  2016-01-01      Friday           1                  0.0                  1.0
1  2016-01-02    Saturday           1                  1.0                  1.0
2  2016-01-03      Sunday           1                  1.0                  0.0
3  2016-01-04      Monday           0                  1.0                  0.0
4  2016-01-05     Tuesday           0                  0.0                  0.0

---
## 4.3 File `air_store_info.csv`


Đối với thông tin của nhà hàng ở đây chúng ta sử dụng phiên bản xử lý trước thay vì phiên bản hiện tại của cuộc thi (vì nó chứa dữ liệu quan trọng của các trạm quan sát thời tiết).

Ta có thể truy cập dữ liệu này ở [đây](https://www.kaggle.com/huntermcgushion/rrv-weather-data).
Dữ liệu này chứa thêm các đặc trưng như: station_id (id của các trạm gần nhà hàng nhất), station_latitude (vĩ độ của trạm), station_longitude (kinh độ),…

In [ ]:
air_store_info = pd.read_csv('data/air_store_info.csv')

air_store_info.head()

air_store_id  air_genre_name  ...   latitude   longitude
0  air_0f0cdeee6c9bf3d7  Italian/French  ...  34.695124  135.197852
1  air_7cc17a324ae5c7dc  Italian/French  ...  34.695124  135.197852
2  air_fee8dcf4d619598e  Italian/French  ...  34.695124  135.197852
3  air_a17f0778617c76e2  Italian/French  ...  34.695124  135.197852
4  air_83db5aff8f50478e  Italian/French  ...  35.658068  139.751599

[5 rows x 5 columns]

In [ ]:
air_store_info = pd.read_csv('data/air_store_info_with_nearest_active_station.csv')

air_store_info.head()

air_store_id  air_genre_name  ... station_vincenty  station_great_circle
0  air_0f0cdeee6c9bf3d7  Italian/French  ...         1.277232              1.274882
1  air_7cc17a324ae5c7dc  Italian/French  ...         1.277232              1.274882
2  air_fee8dcf4d619598e  Italian/French  ...         1.277232              1.274882
3  air_a17f0778617c76e2  Italian/French  ...         1.277232              1.274882
4  air_83db5aff8f50478e  Italian/French  ...         3.730672              3.739835

[5 rows x 12 columns]

---
## 4.4 File `sample_submission.csv`


Tiếp theo chúng ta xử lý tập test, ở đây chúng ta sử dụng file `sample_submission.csv` để lấy dữ liệu test. Ta trích xuất id của nhà hàng và ngày tháng từ cột id. Vì id trong sample là chuỗi gồm id + visit_date.

**Giải pháp :**

-> Ta cần tách ra **id** riêng và **visit_date** riêng.

Cột `test_number` nhằm để các kết quả dự đoán (bài nộp) được sắp xếp theo đúng thứ tự.

In [ ]:
submission = pd.read_csv('data/sample_submission.csv')
submission['air_store_id'] = submission['id'].str.slice(0, 20)
submission['visit_date'] = submission['id'].str.slice(21)
submission['is_test'] = True
submission['visitors'] = np.nan
submission['test_number'] = range(len(submission))

submission.head()

id  visitors  ... is_test test_number
0  air_00a91d42b08b08d9_2017-04-23       NaN  ...    True           0
1  air_00a91d42b08b08d9_2017-04-24       NaN  ...    True           1
2  air_00a91d42b08b08d9_2017-04-25       NaN  ...    True           2
3  air_00a91d42b08b08d9_2017-04-26       NaN  ...    True           3
4  air_00a91d42b08b08d9_2017-04-27       NaN  ...    True           4

[5 rows x 6 columns]

* Sau khi đã tách ra `air_store_id` và `visit_date`, xóa cột `id`

In [ ]:
data = pd.concat((air_visit, submission.drop('id', axis='columns')))

data.head()

air_store_id  visit_date  visitors was_nil is_test  test_number
0  air_00a91d42b08b08d9  2016-07-01      35.0   False     NaN          NaN
1  air_00a91d42b08b08d9  2016-07-02       9.0   False     NaN          NaN
2  air_00a91d42b08b08d9  2016-07-03       0.0    True     NaN          NaN
3  air_00a91d42b08b08d9  2016-07-04      20.0   False     NaN          NaN
4  air_00a91d42b08b08d9  2016-07-05      25.0   False     NaN          NaN

---
## 4.5 Merge dataset


Hợp nhất 2 bộ dữ liệu train và sample_submission sẽ giúp cho quá trình chọn lọc dữ liệu sau này thực hiện đơn giản hơn.


In [ ]:
data['is_test'].fillna(False, inplace=True)
data = pd.merge(left=data, right=date_info, on='visit_date', how='left')
data = pd.merge(left=data, right=air_store_info, on='air_store_id', how='left')
data['visitors'] = data['visitors'].astype(float)

data.head()

air_store_id  visit_date  ...  station_vincenty station_great_circle
0  air_00a91d42b08b08d9  2016-07-01  ...          0.416011             0.415906
1  air_00a91d42b08b08d9  2016-07-02  ...          0.416011             0.415906
2  air_00a91d42b08b08d9  2016-07-03  ...          0.416011             0.415906
3  air_00a91d42b08b08d9  2016-07-04  ...          0.416011             0.415906
4  air_00a91d42b08b08d9  2016-07-05  ...          0.416011             0.415906

[5 rows x 21 columns]

--- 
## 4.6 Dữ liệu thời tiết

Do dữ liệu về thời tiết được lưu trữ ở nhiều file khác nhau, mỗi file ứng với 1 trạm thời tiết. 

**Giải pháp**

* Đầu tiên chúng ta nhập vào dữ liệu thời tiết.
* Liên kết `1663` file thời tiết với nhau vào bảng `weather`. Dữ liệu sau này sẽ chỉ xử lí thông qua bảng này.
* Trong bảng này có nhiều dữ liệu không tồn tại ~ NaN với nhiều lý do:    

  * 1 khu vực quanh năm nắng nóng sẽ không cần tới loại dữ liệu snowfall và tương tự.
  * 1 số trạm thời tiết nhỏ chỉ có 1 nhiệm vụ là đo lượng mưa và có những trạm lớn sẽ đo nhiều thông số thời tiết.
  * 1 số trạm ở khu vực vùng núi sẽ không thể đo đạc các chỉ số liên qua đến biển.

In [ ]:
import glob
weather_dfs = []

for path in glob.glob('data/1-1-16_5-31-17_Weather/*.csv'):
    weather_df = pd.read_csv(path)
    weather_df['station_id'] = path.split('\\')[-1].rstrip('.csv')
    weather_dfs.append(weather_df)


weather = pd.concat(weather_dfs, axis='rows')
weather.rename(columns={'calendar_date': 'visit_date'}, inplace=True)


In [ ]:
weather.head(20)

visit_date  ...                                         station_id
0   2016-01-01  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
1   2016-01-02  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
2   2016-01-03  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
3   2016-01-04  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
4   2016-01-05  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
5   2016-01-06  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
6   2016-01-07  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
7   2016-01-08  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
8   2016-01-09  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
9   2016-01-10  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
10  2016-01-11  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
11  2016-01-12  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
12  2016-01-13  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
13  2016-01-14  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
14  2016-01-15  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
15  2016-01-16  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
16  2016-01-17  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
17  2016-01-18  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
18  2016-01-19  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...
19  2016-01-20  ...  data/1-1-16_5-31-17_Weather/hokkaido_kamikawa_...

[20 rows x 16 columns]

---
## 4.7 Trích chọn dữ liệu thời tiết

Sau khi đã có thông tin về các chỉ số liên quan đến thời tiết, Ta có thể xem xét để trích chọn các giá trị thực sự ảnh hưởng đến số lượng khách hàng.
- Vì là số lượng khách hàng được tính trong 1 ngày nên các giá trị về nhiệt độ cao nhất hay thấp nhất trong 1 ngày ảnh hưởng ít đến số lượng khách.
- Đối với nhà hàng, thời gian mở cửa thường không thay đổi nên các chỉ số về mặt trời mọc hay lặn sẽ không ảnh hưởng đến các quyết định của nhà hàng.
- Ngoài ra còn có một số yếu tố chính, ảnh hưởng trực tiếp các yếu tố khác như nhiệt độ thường ảnh hưởng đến các yếu tố về tuyết nên cần lượt bỏ bớt.

Sau khi đã cân nhắc và xem xét, ta quyết định chọn ra các giá trị quan trọng đến số lượng khách:
- Lượng mưa
- Nhiệt độ
- (lượt bỏ các giá trị khác)

Tiếp theo, tiến hành tiền xử lý các giá trị thiếu bằng cách sử dụng giá trị trung bình (nhiệt độ và lượng mưa) hằng ngày trên **toàn cầu** để thay thế chúng 

In [ ]:
means = weather.groupby('visit_date')[['avg_temperature', 'precipitation']].mean().reset_index()
means.rename(columns={'avg_temperature': 'global_avg_temperature', 'precipitation': 'global_precipitation'}, inplace=True)
weather = pd.merge(left=weather, right=means, on='visit_date', how='left')
weather['avg_temperature'].fillna(weather['global_avg_temperature'], inplace=True)
weather['precipitation'].fillna(weather['global_precipitation'], inplace=True)

weather[['visit_date', 'avg_temperature', 'precipitation']].head()

visit_date  avg_temperature  precipitation
0  2016-01-01         2.868353       0.564662
1  2016-01-02         5.279225       2.341998
2  2016-01-03         6.589978       1.750616
3  2016-01-04         5.857883       1.644946
4  2016-01-05         4.556850       3.193625

---
## 4.8 Kết quả thu được

Sau quá trình khám phá ta thu được bộ dữ liệu như sau:

In [ ]:
data['visit_date'] = pd.to_datetime(data['visit_date'])
data.index = data['visit_date']
#data.sort_values(['air_store_id', 'visit_date'], inplace=True)

data.head()

air_store_id  ... station_great_circle
visit_date                        ...                     
2016-07-01  air_00a91d42b08b08d9  ...             0.415906
2016-07-02  air_00a91d42b08b08d9  ...             0.415906
2016-07-03  air_00a91d42b08b08d9  ...             0.415906
2016-07-04  air_00a91d42b08b08d9  ...             0.415906
2016-07-05  air_00a91d42b08b08d9  ...             0.415906

[5 rows x 21 columns]

# 5. TIỀN XỬ LÝ DỮ LIỆU


##5.1 Lượt bỏ outliers

Ta thực hiện lượt bỏ bớt các phần tử có giá trị lớn nhưng không thể hiện đúng phân bố của dữ liệu gây nhiễu cho quá trình huấn luyện (outlier).


*Ví dụ:* Số lượt truy cập vào dịp năm mới rõ ràng không đại diện cho các thời điểm còn lại trong năm.

Giả sử rằng lượt truy cập tuân theo phân phối chuẩn cho mỗi nhà hàng.

Ta chọn hệ số Z là 2.4 để có khoản tin cậy khoảng 97% và lượt bỏ khoảng 3% số lượng giá trị outlier. Bạn có thể tham khảo thêm tại [đây](https://www.mathsisfun.com/data/confidence-interval-calculator.html) và ở [đây](https://www.statisticshowto.com/probability-and-statistics/z-score/) để biết thêm về hệ số Z.


 <font size="5"> $$ x - \overline{x} = Z * s  $$  </font>

Trong đó: 
- $ \overline{x}: mean $
- $ Z: \text{hệ số score} $
- $ s:  \text{độ lệch chuẩn}$


|   | $$ Z $$ |
|---|---|
| 80% | 1.282|
| 85% | 1.440 | 
| 90% | 1.645   |
| 95% | 1.960  | 
| 95% | 2.576 |
| 99.5% | 2.576 |
| 99.9% | 3.291 |

Những dòng có giá trị outlier sẽ được thay thế bằng max của những giá trị trong khoảng tin cậy.

Tiếp theo ta tính **log1p** để transform lại dữ liệu, giúp dữ liệu phân bố tốt hơn, không bị lệch phải do các giá trị outlier chưa được lượt bỏ hết.

In [ ]:
def find_outliers(series):
    return (series - series.mean()) > 2.4 * series.std()


def cap_values(series):
    outliers = find_outliers(series)
    max_val = series[~outliers].max()
    series[outliers] = max_val
    return series


stores = data.groupby('air_store_id')
data['is_outlier'] = stores.apply(lambda g: find_outliers(g['visitors'])).values
data['visitors_capped'] = stores.apply(lambda g: cap_values(g['visitors'])).values
data['visitors_capped_log1p'] = np.log1p(data['visitors_capped'])

data.head()

air_store_id  ... visitors_capped_log1p
visit_date                        ...                      
2016-07-01  air_00a91d42b08b08d9  ...              3.583519
2016-07-02  air_00a91d42b08b08d9  ...              2.302585
2016-07-03  air_00a91d42b08b08d9  ...              0.000000
2016-07-04  air_00a91d42b08b08d9  ...              3.044522
2016-07-05  air_00a91d42b08b08d9  ...              3.258097

[5 rows x 24 columns]

## 5.2 Trích xuất thêm các đặc trưng.

 Các đặc trưng cần thêm:
 - Ngày cuối tuần(`is_weekend`)
 - Ngày trong tháng(`day_of_month`)

Những đặt trưng này có thể được coi là đại diện cho thời điểm
- Thường hay ăn chơi dịp cuối tuần.
- Mọi người được trả tiền trong tháng, giả sử họ được trả tiền hàng tháng.


In [ ]:
data['is_weekend'] = data['day_of_week'].isin(['Saturday', 'Sunday']).astype(int)
data['day_of_month'] = data['visit_date'].dt.day

## 5.3 Tính trọng số theo cấp số nhân cho số lượng khách hàng.

- Trọng số theo cấp số nhân - Exponentially weighted means (EWM) 

- Ta tính [EWM](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.ewm.html) để tăng trọng số cho số lượng khách hàng của ngày hiện tại dựa trên những ngày trước đó, điều này nhằm mục đích nằm bắt được xu hướng của khách hàng.

- Để tính EWM cần biết tham số **alpha**. Ở đây ta dùng thư viện [optimize](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.differential_evolution.html) để tìm **alpha** tốt nhất cho giá trị EWM. Ta dùng độ lỗi **mean square error** để làm giá trị trả về cho hàm cần tối ưu. Ở đây ta cần tìm `min` của hàm f.

In [ ]:
from scipy import optimize


def calc_shifted_ewm(series, alpha, adjust=True):
    return series.shift().ewm(alpha=alpha, adjust=adjust).mean()


def find_best_signal(series, adjust=False, eps=10e-5):
    
    def f(alpha):
        shifted_ewm = calc_shifted_ewm(series=series, alpha=min(max(alpha, 0), 1), adjust=adjust)
        corr = np.mean(np.power(series - shifted_ewm, 2))
        return corr
     
    res = optimize.differential_evolution(func=f, bounds=[(0 + eps, 1 - eps)])
    
    return calc_shifted_ewm(series=series, alpha=res['x'][0], adjust=adjust)


roll = data.groupby(['air_store_id', 'day_of_week']).apply(lambda g: find_best_signal(g['visitors_capped']))
data['optimized_ewm_by_air_store_id_&_day_of_week'] = roll.sort_index(level=['air_store_id', 'visit_date']).values

roll = data.groupby(['air_store_id', 'is_weekend']).apply(lambda g: find_best_signal(g['visitors_capped']))
data['optimized_ewm_by_air_store_id_&_is_weekend'] = roll.sort_index(level=['air_store_id', 'visit_date']).values

roll = data.groupby(['air_store_id', 'day_of_week']).apply(lambda g: find_best_signal(g['visitors_capped_log1p']))
data['optimized_ewm_log1p_by_air_store_id_&_day_of_week'] = roll.sort_index(level=['air_store_id', 'visit_date']).values

roll = data.groupby(['air_store_id', 'is_weekend']).apply(lambda g: find_best_signal(g['visitors_capped_log1p']))
data['optimized_ewm_log1p_by_air_store_id_&_is_weekend'] = roll.sort_index(level=['air_store_id', 'visit_date']).values

data.head()

air_store_id  ... optimized_ewm_log1p_by_air_store_id_&_is_weekend
visit_date                        ...                                                 
2016-07-01  air_00a91d42b08b08d9  ...                                              NaN
2016-07-02  air_00a91d42b08b08d9  ...                                              NaN
2016-07-03  air_00a91d42b08b08d9  ...                                         2.302585
2016-07-04  air_00a91d42b08b08d9  ...                                         3.583519
2016-07-05  air_00a91d42b08b08d9  ...                                         3.454852

[5 rows x 30 columns]

## 5.4 Tính toán các đặc trưng.

Tính các đặc trưng với tham số **alpha** cố định như:
- Trung bình (mean)
- Trung vị (median)
- Độ lệch chuẩn (standard deviation)
- Số lượng giá trị
- Tối thiểu (min)
- Tối đa (max)
- Giá trị EWM.

In [ ]:
def extract_precedent_statistics(df, on, group_by):

    df.index.name = None
    
    df.sort_values(group_by + ['visit_date'], inplace=True)
    
    groups = df.groupby(group_by, sort=False)
    
    stats = {
        'mean': [],
        'median': [],
        'std': [],
        'count': [],
        'max': [],
        'min': []
    }
    
    exp_alphas = [0.1, 0.25, 0.3, 0.5, 0.75]
    stats.update({'exp_{}_mean'.format(alpha): [] for alpha in exp_alphas})
    
    for _, group in groups:
        
        shift = group[on].shift()
        roll = shift.rolling(window=len(group), min_periods=1)
        
        stats['mean'].extend(roll.mean())
        stats['median'].extend(roll.median())
        stats['std'].extend(roll.std())
        stats['count'].extend(roll.count())
        stats['max'].extend(roll.max())
        stats['min'].extend(roll.min())
        
        for alpha in exp_alphas:
            exp = shift.ewm(alpha=alpha, adjust=False)
            stats['exp_{}_mean'.format(alpha)].extend(exp.mean())
    
    suffix = '_&_'.join(group_by)
    
    for stat_name, values in stats.items():
        df['{}_{}_by_{}'.format(on, stat_name, suffix)] = values


extract_precedent_statistics(
    df=data,
    on='visitors_capped',
    group_by=['air_store_id', 'day_of_week']
)

extract_precedent_statistics(
    df=data,
    on='visitors_capped',
    group_by=['air_store_id', 'is_weekend']
)

extract_precedent_statistics(
    df=data,
    on='visitors_capped',
    group_by=['air_store_id']
)

extract_precedent_statistics(
    df=data,
    on='visitors_capped_log1p',
    group_by=['air_store_id', 'day_of_week']
)

extract_precedent_statistics(
    df=data,
    on='visitors_capped_log1p',
    group_by=['air_store_id', 'is_weekend']
)

extract_precedent_statistics(
    df=data,
    on='visitors_capped_log1p',
    group_by=['air_store_id']
)

data.sort_values(['air_store_id', 'visit_date']).head()


air_store_id  ... visitors_capped_log1p_exp_0.75_mean_by_air_store_id
2016-07-01  air_00a91d42b08b08d9  ...                                                NaN 
2016-07-02  air_00a91d42b08b08d9  ...                                           3.583519 
2016-07-03  air_00a91d42b08b08d9  ...                                           2.622819 
2016-07-04  air_00a91d42b08b08d9  ...                                           0.655705 
2016-07-05  air_00a91d42b08b08d9  ...                                           2.447318 

[5 rows x 96 columns]

## 5.5 Mã hoá dữ liệu

In [ ]:
data = pd.get_dummies(data, columns=['day_of_week', 'air_genre_name'])
data.head()

air_store_id  ... air_genre_name_Yakiniku/Korean food
2016-07-01  air_00a91d42b08b08d9  ...                                   0
2016-07-02  air_00a91d42b08b08d9  ...                                   0
2016-07-03  air_00a91d42b08b08d9  ...                                   0
2016-07-04  air_00a91d42b08b08d9  ...                                   0
2016-07-05  air_00a91d42b08b08d9  ...                                   0

[5 rows x 115 columns]

# 6. XÂY DỰNG MÔ HÌNH


## Chuẩn bị dữ liệu train và test

tổng hợp dữ liệu như sau:
tập train = dữ liệu tổng - dữ liệu test - dữ liệu đánh dấu outlier - dữ liệu null. 

một số dữ liệu được đánh giá là không thể sử dụng trong training sẽ được lược bỏ để cải thiện quá trình. những dữ liệu này sẽ được đưa vào bảng to_drop.
tách biệt dữ liệu x_train ( ma trận của tập huấn luyện ) và dữ liệu y_train ( vector của mô hình huấn luyện).


In [23]:
data['visitors_log1p'] = np.log1p(data['visitors'])
train = data[(data['is_test'] == False) & (data['is_outlier'] == False) & (data['was_nil'] == False)]
test = data[data['is_test']].sort_values('test_number')

to_drop = ['air_store_id', 'is_test', 'test_number', 'visit_date', 'was_nil',
           'is_outlier', 'visitors_capped', 'visitors', 'air_area_name',
           'station_id', 'station_latitude', 'station_longitude', 'station_vincenty',
           'station_great_circle', 'visitors_capped_log1p','latitude_str','longitude_str']
train = train.drop(to_drop, axis='columns')
train = train.dropna()
test = test.drop(to_drop, axis='columns')
X_train = train.drop('visitors_log1p', axis='columns')
X_test = test.drop('visitors_log1p', axis='columns')
y_train = train['visitors_log1p']

In [ ]:
import lightgbm as lgbm
from sklearn import metrics
from sklearn import model_selection


np.random.seed(42)

model = lgbm.LGBMRegressor(
    objective='regression',
    max_depth=5,
    num_leaves=9,
    learning_rate=0.007,
    n_estimators=30000,
    min_child_samples=80,
    subsample=0.8,
    colsample_bytree=1,
    reg_alpha=0,
    reg_lambda=0,
    random_state=np.random.randint(10e6)
)

n_splits = 6
cv = model_selection.KFold(n_splits=n_splits, shuffle=True, random_state=42)

val_scores = [0] * n_splits

sub = submission['id'].to_frame()
sub['visitors'] = 0

feature_importances = pd.DataFrame(index=X_train.columns)

for i, (fit_idx, val_idx) in enumerate(cv.split(X_train, y_train)):
    
    X_fit = X_train.iloc[fit_idx]
    y_fit = y_train.iloc[fit_idx]
    X_val = X_train.iloc[val_idx]
    y_val = y_train.iloc[val_idx]
    
    model.fit(
        X_fit,
        y_fit,
        eval_set=[(X_fit, y_fit), (X_val, y_val)],
        eval_names=('fit', 'val'),
        eval_metric='l2',
        early_stopping_rounds=200,
        feature_name=X_fit.columns.tolist(),
        verbose=False
    )
    
    val_scores[i] = np.sqrt(model.best_score_['val']['l2'])
    sub['visitors'] += model.predict(X_test, num_iteration=model.best_iteration_)
    feature_importances[i] = model.feature_importances_
    
    print('Fold {} RMSLE: {:.5f}'.format(i+1, val_scores[i]))
    
sub['visitors'] /= n_splits
sub['visitors'] = np.expm1(sub['visitors'])

val_mean = np.mean(val_scores)
val_std = np.std(val_scores)

print('Local RMSLE: {:.5f} (±{:.5f})'.format(val_mean, val_std))

Fold 1 RMSLE: 0.56862
Fold 2 RMSLE: 0.57090
Fold 3 RMSLE: 0.56877
Fold 4 RMSLE: 0.56884
Fold 5 RMSLE: 0.57210
Fold 6 RMSLE: 0.57027
Local RMSLE: 0.56992 (±0.00129)


# 7. ĐÁNH GIÁ MÔ HÌNH 